### ISSUE TITLE
Tracking Issue for `normalize_lexically`

### ISSUE THREAD

**Issue by ChrisDenton** [rust-lang/rust#134694]
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/issues/134694)*

Feature gate: `#![feature(normalize_lexically)]`

This is a tracking issue for `Path::normalize_lexically` that is used to normalize a path, including `..` parent references, without touching the filesystem.

### Public API

```rust
// std::path

impl Path {
    pub fn normalize_lexically(&self) -> Result<PathBuf, NormalizeError> {
}
```

### Steps / History

- [x] [API Change Proposal](https://github.com/rust-lang/libs-team/issues/396)
- [x] Implementation: #134696
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- Should `PathBuf` have a method that normalizes in-place? Normalizing a path just removes components.
- Should this disallow resolving to the empty path or the single `.` path.
- Should this only allow paths without a root or prefix? It may be surprising that `base_path.join(user_path.normalize_lexically()?)` can still escape the base path.

The last two can be summarised as: should this strictly require paths to be sub-paths?

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by clarfonthey** [comment]
How would you do a function on `Path` that removes components of the path in-place, given what I mentioned?

**Comment by ChrisDenton** [comment]
No, I meant have a function on Path that returns a PathBuf and a separate function on PathBuf that does in-place normalisation.

**Comment by MikkelPaulson** [comment]
Thanks for your concurrence. I have to disagree about in-place operation, though. Very few of the existing in-place examples affect the size in memory, with `String::replace_range` being one exception. Otherwise, they're mainly things like `[T]::sort()` and `str::make_ascii_uppercase()` that can operate directly on smart pointers because the lengths remain the same (and are valuable for that exact reason). Note that `str` has `to_ascii_uppercase(&self) -> String` and `make_ascii_uppercase(&mut self)`, but only `to_uppercase(&self) -> String` (with no `String::make_uppercase(&mut self)`).

---

The `Result` behaviour seems to be quite a considered choice, from the following comment by @Amanieu on https://github.com/rust-lang/libs-team/issues/396:

> We [@rust-lang/libs-api] recognize that the behavior of erroring differs from other languages, but we believe that this behavior is more useful in practice for path validation.

I still disagree with the team's perspective on this. *Validation* and *normalization* are separate goals, and using this for validation is both overkill and still falls short of being a one-stop validation check. I predict a lot of

```rust
if path.normalize_lexically().is_ok_and(|p| p.is_relative()) {
    // happy path
```

Which is both a potential footgun (easy to forget the `p.is_relative()`) and overkill (allocating and then throwing the value away). Making this infallible will make it clear that it is not to be used for validation. I propose instead adding `fn is_bounded(&self) -> bool`, which can be trivially implemented without allocating:

```rust
fn is_bounded(&self) -> bool {
    use std::path::Component::*;

    self.components()
        .try_fold(0usize, |depth, component| match component {
            Prefix(_) | RootDir => None,
            CurDir => Some(depth),
            Normal(_) => Some(depth + 1),
            ParentDir => depth.checked_sub(1),
        })
        .is_some()
}
```

**Comment by MikkelPaulson** [comment]
It seems like there's a case to be made for ripping the band-aid and overhauling the way [edge cases](https://github.com/rust-lang/rust/issues/148267) are handled, but if so that's more of a side quest for the `trim_trailing_sep` issue than a matter for here and now.

The other question that is maybe a not-now problem is if adding lexical treatment of paths is decoupling us from *filesystem* paths, where we might need to handle paths in a platform-independent (but still not necessarily UTF-8–compliant) way.

---

Having slept on it, I think the open questions in the issue description are open because the problem being solved here is not well defined. It's a bit validation and a bit sanitization and doesn't really succeed at either.

First, `normalize_lexically` will allocate (if not all the time if we use `Cow`), so it should only be used in a case where the normalized result will be *used*. There are two validation cases that I can anticipate where the result will be ultimately thrown away:

```rust
// Ensure that the path doesn't escape the pwd
if path.normalize_lexically().is_ok_and(|p| p.is_relative()) { ... }

// Ensure that the path doesn't attempt any funny business (even abc/../def)
if path.normalize_lexically() == Ok(path) && path.is_relative() { ... }
```

I think it's worth covering these use cases with separate non-allocating function calls, for which I suggest `is_lexically_bounded()` and `is_normalized()`. (Once normalized, there's no lexical/filesystem angle to "is it weirdly constructed?", so `is_lexically_normalized()` isn't necessary. The lexical aspect only registers in how that normalization happens. However, one could maybe add `is_bounded()` that *does* follow symlinks.)

Conveniently, `is_normalized()` can be reused in `normalize_lexically()` to identify when `Cow::Borrowed` can be used vs. when `Cow::Owned` is needed.

I've already implemented as part of an extension trait I'm working on, so when I get a chance (probably tomorrow) I'll put up a draft PR with the logic ported to `std` for discussion purposes.

---

One other thought about the current implementation: `"abc/.."` should normalize to `"."` not `""`, because `fs::read_dir("abc/..")` is valid (if `abc` exists) but `fs::read_dir("")` is not.

**Comment by ChrisDenton** [comment]
> Definitely agree that `""` is incorrect for `abc/..`, and it should be `.`. An empty path is pretty much never desired in any case except initialization, so, we should try to avoid it unless one is passed as input.

I would strongly agree with that.

**Comment by MikkelPaulson** [comment]
@ChrisDenton @clarfonthey Pull request is up: #152740. Took me longer than expected to figure out a way to get `Components` to admit when it's silently ignoring repeated separators and `.`, which was needed for `is_normalized`. Notably, I made the call that a `Path` must have its trailing slash stripped to be normalized, because doing so means that there's a 1:1 relationship between normalized paths and sequences of `Component`s. I'm not wedded to that idea, though.

I went a bit overboard with doctests to illustrate how I approached edge cases. I haven't tested non-Unix environments at all, unfortunate because Windows paths are more complex.

**Comment by MikkelPaulson** [comment]
@ChrisDenton So do I understand your position to be that `normalize("abc/")` should produce `"abc/"`? Likewise `"abc/." => "abc"` and `"abc/./" => "abc/"`? I'm also not opposed to this so long as we're explicit about it.

Did you have the chance to look over the [draft PR](https://github.com/rust-lang/rust/pull/152740) for `is_lexically_bounded()`/`is_normalized()`? I'm not expecting a detailed review, but agreement in principle would be appreciated before advancing it for libs-api review.

**Comment by danieltrick** [comment]
> The last two can be summarised as: should this strictly require paths to be sub-paths?

For _relative_ paths it definitely **must** be allowed to have paths that reach "above" the current directory.

At the very least, we need an option that allows this.

Suppose we are writing a command-line utility and the user calls the program like this, which is a 100% legitimate use case:
```sh
$ ./program ../../some/path/file.txt
```

A given path like this must be preserved, at least optionally, but should still be normalized properly:
* `../../some/./path/file.txt` ⭢ `../../some/path/file.txt`
* `foo/bar/../../../some/./path/file.txt` ⭢ `../some/path/file.txt`

-----

I understand that there are other use-cases that need to ensure that the path can't escape the "base" directory.

Web-server with an HTML "root" directory would be an example.

So, I suggest extending the API like this:
```rust
impl Path {
    pub fn normalize_lexically(&self, require_subpath: bool) -> Result<PathBuf, NormalizeError>
}
```

**Comment by Qix-** [comment]
/bikeshed

Not a fan of boolean flags. Normalize means to normalize, there's nothing "ab"-normal about a path that traverses into the parent, especially since the only reason a `NormalizeError` would occur according to the PR.

I'd rather see something like:

```rust
/// Normalizes a path lexically.
///
/// Returns `Ok(path)` if the path lexically normalized to a subpath (i.e. `foo`, `foo/bar`, etc).
///
/// Returns `Err(path)` if the path lexically normalized to a path that traverses upward
/// (i.e. `../foo`, `../../foo/bar`, etc).
///
/// Users of the `Err(path)` value should be aware of potential
/// [path traversal attacks](https://owasp.org/www-community/attacks/Path_Traversal)
/// by using paths that traverse upward.
fn normalize(&self) -> Result<PathBuf, PathBuf>;
```

**Comment by schneems** [comment]
I think this is a big gap in the current `std::fs`. With `absolute` you can ensure you're not dealing with relative paths, but it doesn't do anything with `..`. With `canonicalize,` you get rid of `..` but also expand things like `/tmp` to forms that might be surprising to people. An API like this one helps people show what the actual target is without following symlinks, which is nice. Thank you, @ChrisDenton!

I don't like @Qix- suggestion of returning `Result<PathBuf, PathBuf>`. I view a `Result` as either: "here's the valid data you asked for" or "here's the data saying why we couldn't give you what you wanted," and from a type standpoint, `Result<T, T>` gives you back something you already had when it fails with no (type) way to distinguish between success and failure.

- If there's only one failure mode, possibly `Option<PathBuf>` says "we could normalize this or we couldn't," and you can annotate what "could not normalize means" in the docs.
- I'm also okay with the current error behavior for `/a/..`, but maybe provide an example of how to map an Err into a root path (possibly make this easier) like providing a `Into<PathBuf>` or `NormalizeError.to_root()`.

Either way, I suggest we add some notes about common platform behaviors. I mistakenly believed the docs meant that OSX would err if you tried to escape root, rather than it treating `/a/../..` as (roughly) the same as root. We can link to [POSIX behavior](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_13) and state the current behaviors with some asterisks that it's not guaranteed behavior. I did something like that in this PR https://github.com/rust-lang/rust/pull/158574. 

I would also like to see some examples like:
- normalizing an absolute path with an `..`
- normalizing a path that starts as a relative path like `../../a/b` (seems like these errors and we should be able to distinguish between that and `/a/../..`) and/or suggest chaining this with `absolute()`.

**Comment by Qix-** [comment]
@schneems Then a proper `enum` could be used to be clearer about what exactly happened.

```rust
enum NormalizeError {
    TraverseAboveRoot
}

enum Normalized {
    /// The relative path fit cleanly into the subpath; it did not traverse upward out of the current path.
    Subpath(PathBuf),
    /// The relative path resolved to a path above the current path.
    Superpath(PathBuf),
    /// The normalized path is an absolute path.
    Absolute(PathBuf),
}

impl Normalized {
    fn into_inner(self) -> PathBuf {
        match self {
            Self::Subpath(p) => p,
            Self::Superpath(p) => p,
            Self::Absolute(p) => p,
        }
    }
}

fn normalize(&self) -> Result<Normalized, NormalizeError>;
```

**Comment by schneems** [comment]
TLDR, I think we can get away with a two-element enum and dropping the Result, making the return infallible.

Riffing on your design, I think we need to represent when an absolute path had an escape attempt. Then the exhaustive cases are:

```rust
enum Normalized {
    /// The relative path fit cleanly into the subpath; it did not traverse upward out of the current path.
    Relative(PathBuf),
    /// The relative path resolved to a path above the current path.
    RelativeEscaped(PathBuf, PathBuf), // (escaped, normalized)
    /// The absolute path normalized cleanly with no escape attempt
    Absolute(PathBuf),
    /// The absolute path had an escape attempt, the remainder is available
    AbsoluteEscaped(PathBuf, PathBuf), // (escaped, normalized)
}
```

The relative/absolute information is encoded in the `PathBuf` already, so we could collapse to:

```rust
enum Normalized {
    Ok(PathBuf),
    Escaped(PathBuf, PathBuf) // (escaped, normalized)
}
```

Possibly have some accessors for convenience:

```rust
impl Normalized {
    /// Returns Some when the input path was absolute. When an escape attempt
    /// was made, the attempt is truncated at the filesystem root. `/` for unix and 
    /// `<drive>:\` (i.e. `C:\`) for Windows. The return is guaranteed to be an absolute path
    /// May return root, will never be empty.
    fn absolute(&self) -> Option<&Path>  //...

    /// Returns a normalized path if the target was relative and did not escape root.
    /// return is guaranteed to be relative. May return an empty path
    fn relative(&self) -> Option<&Path> // ...

    /// Returns only if no escape attempt was made or path is absolute
    /// return could be relative or absolute. Return may be empty or root
    fn ok_or_absolute -> Option<&Path>
}
```

Then you could achieve my desired behavior pretty cleanly:

```rust
let normalized = normalize(&PathBuf::from("/tmp/../lol"))
    .absolute();

assert_eq!(
    Some(PathBuf::from("/lol")), 
    normalized
);
```

Behaves similarly to Ruby's `File.expand_path`, or Python's `os.path.normpath` https://docs.python.org/3/library/os.path.html#os.path.normpath:

```
irb(main):003> File.expand_path("/tmp/../lol")
=> "/lol"
```

**Comment by Qix-** [comment]
What would be the two values for `AbsoluteEscaped` for a normalized path after `/foo/../../bar`? That's not entirely clear to me.

Otherwise that'd be fine; technically most utilities allow traversing 'above' root, and just truncate the traversal to `/`, so in cases where you really don't care and just want A Path to be normalized, that would be the escape hatch. I struggle to think of cases where other errors would arise anyway.

**Comment by schneems** [comment]
> What would be the two values for AbsoluteEscaped for a normalized path after /foo/../../bar? That's not entirely clear to me.

I'm thinking about it like this (ruby):

```ruby
path = "/foo/../../bar"
normalized = File.expand_path(path)
# => "/bar"
escaped = path.delete_suffix(normalized)
# => "/foo/../.."
```

So the two values would be 

```rust
Normalized::AbsoluteEscaped(
    PathBuf::new("/foo/../.."), // everything that "escaped" root (truncated at root)
    PathBuf::new("/bar"), // normalized, will never contain `..` or `.`
)
```

They're ordered such that the part that is truncated to the root comes first.

Earlier when i said:

> I think we need to represent when an absolute path had an escape attempt. 

I don't think it's **strictly** necessary. As you pointed out, most utilities truncate or throw away the data. But for those that do care, it's there. And providing it collapses 3 success and 1 error case into 2 cases. My only worry would be that a tuple `PathBuf, PathBuf` isn't terribly helpful for saying which-is-which. A struct-variant could help, but would be a style departure from the rest of std.

**Comment by Qix-** [comment]
But that breaks with `/foo/../../bar/baz/../qux`, which should show `/bar/qux` and `/foo/../..` in your case, but `delete_suffix('/bar/qux')` would not do anything and shows `/foo/../../bar/baz/../qux` as the "leftover" part. I don't see how that actually tells me anything useful.

**Comment by Skgland** [comment]
> Normalized::AbsoluteEscaped(
>     PathBuf::new("/foo/../.."), // everything that "escaped" root (truncated at root)
>     PathBuf::new("/bar"), // normalized, will never contain `..` or `.`
> )

I find the `truncated at root` confusing as the path `/foo/../..` would ascend one above root so it isn't actually truncated at root.

Instead I would have expected `/foo/../../bar` to result in `/../bar`.
The initial `foo` and the first `..` cancle out leaving `/../bar` the remaining `..` cannot be normalized away as it would ascend beyond root and `bar` cannot cancle it out either.

Some more complex example 
- `"/foo/../../bar/zod/../baz"` would result in `"/../bar/baz"` and `"/bar/baz"`.
- `"/foo/../../bar/zod/../../../baz"` would result in `"/../../baz"` and `"/baz"`.

The important part being how far beyond root a path ascends and into what path it descends from there.
The second path that never contains `..` or `.` would be just the first but with all `..` components removed,
basically ascending into the parent saturating at the root.

> The relative/absolute information is encoded in the `PathBuf` already, so we could collapse to:
> 
> enum Normalized {
>     Ok(PathBuf),
>     Escaped(PathBuf, PathBuf) // (escaped, normalized)
> }

Yeah, I don't think the absolute/relative distinction is helpful here as normalization shouldn't change whether a path is absolute/relative and it can be easily checked with `Path::is_absolute`/`Path::is_relative`.

Further I wouldn't consider a relative path normalizing to a path that ascends beyond `./` to be a case that should be non `Ok` such a path would always begin with a `..` or `c:` (windows drive letter) component.

In summary the only special/error case in my opinion should be absolute paths that ascend beyond root.
And for those there are two ways to normalize them 1. keep .. that ascend too far 2. ignore/drop .. that ascend too far

BTW shouldn't this discussion be moved to e.g. zulip rather than being held on a tracking issue?

**Comment by Qix-** [comment]
But `/../bar/baz` is also meaningless; it's 1) not normalized, 2) has no 'well defined' behavior aside from collapsing to `/bar/baz`.

We have to look at defacto behavior; go to your shell and type `cd /; cd ../../../`. You'll end up in root. It's baked into the linux kernel. Historically, `.` and `..` were real directory entries on the disk; `..` pointed upward to the parent directory inode. For `/..` (again, a real entry), it pointed at `/`'s inode. It's also how utilities find root; something like `get_inode('.') == get_inode('..')`. Otherwise path traversal would have to have error paths for handling `".." not found`.

So this is established and guaranteed behavior. Normalization shouldn't try to sidestep that; there are real usecases for collapsing "above root" to just "root".

**Comment by ChrisDenton** [comment]
> BTW shouldn't this discussion be moved to e.g. zulip rather than being held on a tracking issue?

It would probably be best to create a new issue and link it from here if someone can write up a concrete proposal or explain the competing ideas. Then that can be nominated for libs-api to discuss.

**Comment by schneems** [comment]
Seems good. Just wanted to clarify

> delete_suffix('/bar/qux') would not do anything and shows

Im not literally suggesting using that Ruby code its more like pseudo code. For "/foo/../../bar/baz/../qux" I would expect the two values to be "/foo/../.." (could not remove ".." or ".") and "/bar/qux" (can remove ".." and ".")

Though, as long as I can reliably construct "/bar/qux" from that input I'm happy.

> So this is established and guaranteed behavior. Normalization shouldn't try to sidestep that; there are real usecases for collapsing "above root" to just "root".

The posix spec leaves it at MAY truncate to root. So it's not a guarantee. However, we can stabilize the behavior of the function with a comment and note that platforms may diverge. This lets us optimize for the most common use case. I'm okay saying "any absolute path returns Ok and we truncate at root.” I could make a PR to change behavior without changing data and we could discuss there.

**Comment by lo48576** [comment]
Just out of curiosity, what would be an expected use case of normalization NOT truncating at root?

For truncating at root (i.e., squashing "above the root" to the root), I can easily imagine that the sandboxing-like stuff will want it and it would be important from the security perspective. However, I cannot tell immediately when not-truncating-at-root lexical normalization (i.e., preserving "above the root" levels) would be the best. Maybe "relativization over some base path" or something like that?

If both are useful and use cases would be separate, it might be better to give them different names... though I don't have good idea of such names.

**Comment by schneems** [comment]
I have a few proposed changes:

- https://github.com/rust-lang/rust/pull/159874
- https://github.com/rust-lang/rust/pull/159863
- https://github.com/rust-lang/rust/pull/159866 (left in draft, stacked on 159863. Might help handle @danieltrick's prior [comment](https://github.com/rust-lang/rust/issues/134694#issuecomment-4556586875))

If the implementations are no-bueno, that's fine. I can discuss the abstract idea independently of the implementation details. If you prefer, I can spin up libs-api discussions. Sometimes implementation details (what's easy, what's possible) help to inform my decisions.

**Comment by ChrisDenton** [comment]
This needs libs-api to approve. At the very least it would be good to give an overview of the changes and the motivation in one place.

**Comment by retep998** [comment]
I'd also still like to see some mention of whether this API is intended to normalize forward slash to back slashes on Windows.

**Comment by ChrisDenton** [comment]
> I'd also still like to see some mention of whether this API is intended to normalize forward slash to back slashes on Windows.

The PR above adds documentation for that case.

**Comment by ChrisDenton** [comment]
Concrete proposal:

`normalize_lexically` should always succeed (except perhaps in the empty case?) but return an `enum`. Or a struct that wraps a `LexicalKind` and a `PathBuf`.

## Motivation

There are a variety of different uses for `normalize_lexically`. There are also a variety of different security scenarios that people may (or may not) want to handle whether or not a path is lexically normalized. These are two different properties that should be composable but should also be usable separately. The current design of `normalize_lexically` intertwines them and doesn't let you untangle them.

## Quick sketch

```rust
enum LexicalPath {
    EscapedRoot(PathBuf),
    CurrentDir(PathBuf),
    Absolute(PathBuf),
}
impl LexicalPath {
    fn into_path(self) -> PathBuf {
        match self {
            EscapedRoot(p) | CurrentDir(p) | Absolute(p) => p
        }
    }
}
```

Then people who need to know the details can match on the enum and people who don't have security or other requirements can just use `normalize_lexically().into_path()` (or just `.into()`). It's not (necessarily) an error if a path escapes the root or is an absolute path. The standard library can help but it's ultimately up to the user to judge the situation.

This also suggests to me we should normalize the legacy Windows roots (e.g. `C:path` and `\path`) into their absolute form. Or else we need an `Other` category.

**Comment by eirnym** [comment]
I also wore to always succeed if path can be opened. I'm not sure about enum as proposed above. This sounds more like "classify my path", which is not direct responsibility of this function. 

For example, relative paths "../" and "c:../dir" can be opened, so function must succeed.

**Comment by schneems** [comment]
> For example, relative paths "../" and "c:../dir" can be opened, so function must succeed.

I think the limitation of the function (as implied by "lexically") is that it doesn't perform any disk lookups. i.e. it won't ever produce a `std::io::Error`. Like calling `std::fs::absolute` would if the call to CWD fails. Without knowing CWD, you cannot produce a meaningful normalization of `../` since we have no way to know what is being eaten/folded.

To the concrete proposal: There are cases where we cannot produce a meaninfully different output than what we're given like `../` for those cases I would prefer to hint "nothing changed" or "could not normalize". Versus if we give them an enum, they have to do the mental math to figure out which variants hold normalized outputs and which don't. I think more than a few lazy programmers might just reach for the `Into<PathBuf>` and completely miss that their input couldn't be normalized. So, I prefer to differentiate them with an Err.

Zooming out: I sat with the core of "what exactly is normalization" and wrote some thoughts on a few different ways to think of it, as well as a straw-man proposal. https://gist.github.com/schneems/e0f484bef06ba452199b3eed776203a3. 

> Applying my proposal, I would suggest these for relative paths:
> 
> * Empty input `""` → `"."`
> * CurDir input `"."` → `"."`
> * Eat everything `"a/.."` → `"."`
> * Plain relative `"a"` → `"./a"`
> * Dot relative `"./a"` → `"./a"`
> * Starting dot slash `./a/b` → `./a/b`
> * Slash no dot `a/b` → `./a/b`
> * Relative escape `”.."` → Err
> * Relative escape `”../anything/else"` → Err
> 
> So a normalized relative path ALWAYS starts with a `.` and never ends with one (`Path` strips those already).
> 
> Absolute edge cases:
> 
> * Root `"/"` → `"/"`
> * Absolute eat everything `"/a/.."` → `"/"`
> * Absolute escape `"/../.."` → `"/"` (pin to root)
> * Absolute escape multiple `"/../../../a"` → `"/a"` (pin to root)

I'm also now in Zulip, under "Richard Schneeman" if you want to peel off a part of that and have more of a chat.

**Comment by eirnym** [comment]
There was no intention to make this API to touch filesystem at all. 

My point is that lexical normalisation is a part of link traversal, but not all of it. And it's better to move this responsibility to the caller as all situations might be different. 

E.g. `..` in one context could mean "went beyond safety boundary, in other case it's just a path on a filesystem. 

If there's an intention to have such api, there's a method of `Path` to check if it's a relative path which fully covers the second safety part in their interpretation.

The only path which is in a gray area is empty path which may or may not be considered as equivalent of `.`.

**Comment by schneems** [comment]
@eirnym i still guess I don't totally follow. If you want to represent the traversal there's `std::fs::canonicalize`. 

> E.g. .. in one context could mean "went beyond safety boundary, in other case it's just a path on a filesystem.

Really, the safety provided by lexical normailzation is for containment is limited as even a normalized path could have a symlink that points to a prior path. I still think it's valuable in an application to reduce/remove `..` as it makes reasoning about the logic easier. It reduces the entropy. But it can also cut the other way i.e. a `..` can "escape" a symlink such that folding it lexically does not produce an equivalent path as a physical traversal. Ruby's `File.expand_path` both follows physical links and folds as it goes. I think that's safer but I think there's some subtle dragons there too as you can wind up in situations where the traversal stops due to a problem like a broken symlink. 

Most languages seem to have a number of slightly different APIs with slightly different behavior.  That's why I'm suggesting we try to lock down what "is" and isn't the job of this current feature. That way we can have strong confidence when a PR comes in to accept or reject it. And if someone needs a thing that isn't provided here, it would be a hint to start the process over with a new API with a new stated purpose and goal. I don't think we can have a single API make everyone 100% happy for all their needs.

> The only path which is in a gray area is empty path which may or may not be considered as equivalent of ..

There's a few ambiguous cases in the existing implementation beyond this one (spelled out some differences in a few of my links). I think it's more useful to define this behavior and declare it produces something other programs can use.

**Comment by eirnym** [comment]
The only think I'd like that lexical normalisation won't think for me in what context I'd like to use it. I'd like that relative paths were valid, even they follow above current folder.

**Comment by Skgland** [comment]
While most of the choices seam reasonable/acceptable I have a few notes.

## Absolute Escape 

If lexical normalization should return an Err for any path it should be ̀the absolute root escapes, as that can't be check easily before or after.

I.e. 

> - Absolute escape "/../.." → "/" (pin to root)
> - Absolute escape multiple "/../../../a" → "/a" (pin to root)

should result in Err.

## Relative Escape
For me a relative path can fail to be normalized .. is perfectly fine at the beginning of a normalized path.
So rather than

> So a normalized relative path ALWAYS starts with a . and …

> A normalized relative path either starts with either a "." or at least one ".." .
> "." may only appear as the first component (after an optional prefix)
> ".." may only appear as the first component (after an optional prefix) or after itself

I.e. 

> Relative escape ”.." → Err
> Relative escape ”../anything/else" → Err

shouldn't result in Err. If someone wants to reject these checking for .. after normalization can be done easily.
Though if you insist that these should Err as long as the Err provides the what you call a partially normalized path I could live with that.

## Plain relative

I don't think it is good that we deviate here from all other referenced implementation

> - Plain relative "a" → "./a"

Though I don't feel strongly about it as it should be easy to check for so that one can skip normalization in that case.

## Paths with Prefix
The examples are missing e.g. windows paths with prefix.
Basically I think it should act as 
    1. if the path has a prefix remove it
    2. perform normalization as described before
    3. if 1. removed a prefix add it back

**Comment by ChrisDenton** [comment]
I've started [a discussion with the libs team on zulip](https://rust-lang.zulipchat.com/#narrow/channel/219381-t-libs/topic/normalize_lexically/with/623616683). I'd like to keep that discussion focused on the broad objective of this method rather than the specific details. Without some agreement on what `normalize_lexically` is for we're unlikely to make progress.

**Comment by clarfonthey** [comment]
Nomination for meeting discussion, with nomination shifting from #159863 to this.

**Comment by clarfonthey** [comment]
So, collecting feedback from libs meeting:

* We are in agreement that this should ignore symlinks, and that `a/b/..` is always `a`.
* We are in agreement that there are no parents of the root, so, `/..` is just `/`. (+ for other prefixes as well, noting that `C:..` is not `C:\`)
* We know that this is not enough for security, and that you should not rely solely on this for security reasons. We could potentially find a resource to link here detailing some of the potential issues and why this isn't enough.

We haven't figured out trailing slashes, yet, and it feels like this might be a place where we want to have extra options, or just not trim the trailing separator and let users do it themselves.

**Comment by ChrisDenton** [comment]
Just to be explicit, the feeling in the libs meeting is that this function should just normalize the path and not do anything else, like restrict it to the root. E.g. `a/../..` should be `../`.

As clarafon said, security of paths is a much more complex issue and we can't handle them all in one function. It's also not specific to lexical normalization. This kind of normalization may be one component of securing paths but it may also not be. Even in the case that it is part of securing paths, this one function can't do everything people want without growing a lot of options so it's better for it to do one thing so it can then be composable.
